src/main/java/net/minestom/server/inventory/Inventory.java

```diff
@@ -3,392 +3,194 @@
-    }
-
-    @Override
-    public synchronized void clear() {
-        this.cursorPlayersItem.clear();
-        super.clear();
-    }
+    byte getWindowId();
 
     /**
-     * Refreshes the inventory for all viewers.
+     * Gets the cursor item of a viewer.
+     *
+     * @param player the player to get the cursor item from
+     * @return the player cursor item, air item if the player is not a viewer
      */
-    @Override
-    public void update() {
-        this.viewers.forEach(p -> p.sendPacket(createNewWindowItemsPacket(p)));
-    }
+    @NotNull ItemStack getCursorItem(@NotNull Player player);
 
     /**
-     * Refreshes the inventory for a specific viewer.
-     * <p>
-     * The player needs to be a viewer, otherwise nothing is sent.
+     * Changes the cursor item of a viewer,
+     * does nothing if <code>player</code> is not a viewer.
      *
-     * @param player the player to update the inventory
+     * @param player     the player to change the cursor item
+     * @param cursorItem the new player cursor item
      */
-    public void update(@NotNull Player player) {
-        if (!isViewer(player)) return;
-        player.sendPacket(createNewWindowItemsPacket(player));
-    }
-
-    @Override
-    public @NotNull Set<Player> getViewers() {
-        return unmodifiableViewers;
-    }
+    void setCursorItem(@NotNull Player player, @NotNull ItemStack cursorItem);
 
     /**
-     * This will not open the inventory for {@code player}, use {@link Player#openInventory(Inventory)}.
+     * Handles the provided click from the given player, returning the results after it is applied. If the results are
+     * null, this indicates that the click was cancelled or was otherwise not processed.
      *
-     * @param player the viewer to add
-     * @return true if the player has successfully been added
+     * @param player the player that clicked
+     * @param clickInfo the information about the player's click
+     * @return the results of the click, or null if the click was cancelled or otherwise was not handled
      */
-    @Override
-    public boolean addViewer(@NotNull Player player) {
-        final boolean result = this.viewers.add(player);
-        update(player);
-        return result;
-    }
+    @Nullable ClickResult handleClick(@NotNull Player player, @NotNull ClickInfo clickInfo);
 
     /**
-     * This will not close the inventory for {@code player}, use {@link Player#closeInventory()}.
+     * Gets the click preprocessor for this inventory.
      *
-     * @param player the viewer to remove
-     * @return true if the player has successfully been removed
+     * @return the click preprocessor
      */
-    @Override
-    public boolean removeViewer(@NotNull Player player) {
-        final boolean result = this.viewers.remove(player);
-        setCursorItem(player, ItemStack.AIR);
-        this.clickProcessor.clearCache(player);
-        return result;
-    }
+    @NotNull ClickPreprocessor getClickPreprocessor();
```

Comment:
@ApiStatus.Internal ?

Reply:
Actually on second thought maybe not even included here? Whats the use case for using it directly?

Reply:
this is used in `WindowListener#clickWindowListener` so if it's contained on each inventory it probably should be public and @ApiStatus.Internal, yeah. And it's a field because it uses the inventory size in its logic.

LMK if that helps with anything. I'm not super satisfied with its usage here either, but idk what else to do.